docs/src/proposals/interest-bearing-tokens.md

```diff
@@ -0,0 +1,414 @@
+holding account.
+
+Since interest is accrued continuously, however, just 1 second later, the total
+supply has again increased. A 10% annualized rate gives us a continuous interest
+rate of ~9.5% (`ln(1.1)`), which allows us to calculate the new instantaneous supply.
+
+```
+new_supply = previous_supply * exp(continuous_rate / seconds_per_year)
+new_supply = 1100 * exp(0.095 / 31536000)
+new_supply = 1100.0000033...
+```
+
+Since interest-bearing tokens are constantly accruing interest, they do not have the same
+properties as normal tokens. Most importantly, interest-bearing tokens held in two different
+accounts are only fungible if they have been updated to the same time. For example,
+10 interest-bearing tokens in the year 1999 are worth much more than 10 interest-bearing tokens in the year 3000, since
+those tokens from 1999 have over 1,000 years of interest (see Futurama). Because
+of this, we have to compound interest before any minting, burning, and
+transferring.
+
+### Initial Attempt
+
+The concept seems simple to implement. We can create an SPL token, allow
+tokens to be minted by a new "interest accruing" program, and then make sure to
+call it regularly.
+
+This solution has a few problems. First, how do we compound interest regularly?
+We can create a separate service which periodically sends transactions to update
+all of the accounts. Unfortunately, this will cost the service a lot, especially if
+there are thousands of accounts to update.
+
+We can enforce compounding before any transfer / mint / burn, but then all programs
+and web consumers need to change to make these update calls. How do we change programs
+and wallets to properly update accounts without creating a big `if is_i_token` case
+all over the code?
+
+This solution creates friction for almost all SPL token users: on-chain programs,
+wallets, applications, and end-users.
+
+## Proposed Solution
+
+Following the approach taken by Aave, we can lean on a crucial feature of
+interest-bearing tokens: although the holding's token amount is constantly
+changing, each token holder's proportion of total supply stays the same.
+
+Therefore, we can satisfy the fungibility requirement for tokens on proportional
+ownership. The token interface stays exactly the same, operating on shares instead
+of tokens, and the concept of "token amount" is actually the "UI amount" to be
+displayed by frontends.
+
+Let's go through an example. Imagine we have an interest-bearing token mint with a total of
+10,000 proportional shares, a 20% annual interest rate, and 1 share is currently
+worth 100 tokens. If I transfer 1,000 tokens today, the program converts that
+using the interest-bearing token program and asks it to move
+10 shares. One year later, 1 share is worth 120 tokens, so if I transfer
+1,000 tokens, the program converts and asks to move 8.33 shares. The same concept
+applies to mints and burns.
+
+Following this model, we never have to worry about compounding interest. We
```

Comment:
Would the interest be compounded periodically by the runtime and stored as a number in the i-token? Or would it need to be computed every time by `AmountToUiAmount` and not stored?

Reply:
The runtime would not be involved at all for this, so it would need to be computed every time and not stored anywhere.